test/common/config/subscription_factory_test.cc

```diff
@@ -60,21 +60,124 @@ TEST_F(SubscriptionFactoryTest, NoConfigSpecifier) {
       "Missing config source specifier in envoy::api::v2::core::ConfigSource");
 }
 
-TEST_F(SubscriptionFactoryTest, WrongClusterNameLength) {
+TEST_F(SubscriptionFactoryTest, RestClusterEmpty) {
   envoy::api::v2::core::ConfigSource config;
+  Upstream::ClusterManager::ClusterInfoMap cluster_map;
+
   config.mutable_api_config_source()->set_api_type(envoy::api::v2::core::ApiConfigSource::REST);
-  EXPECT_CALL(cm_, clusters());
+
+  EXPECT_CALL(cm_, clusters()).WillOnce(Return(cluster_map));
   EXPECT_THROW_WITH_MESSAGE(
       subscriptionFromConfigSource(config), EnvoyException,
       "envoy::api::v2::core::ConfigSource must have a singleton cluster name specified");
-  config.mutable_api_config_source()->add_cluster_names("foo");
-  config.mutable_api_config_source()->add_cluster_names("bar");
-  EXPECT_CALL(cm_, clusters());
+}
+
+TEST_F(SubscriptionFactoryTest, GrpcClusterEmpty) {
+  envoy::api::v2::core::ConfigSource config;
+  Upstream::ClusterManager::ClusterInfoMap cluster_map;
+
+  config.mutable_api_config_source()->set_api_type(envoy::api::v2::core::ApiConfigSource::GRPC);
+
+  EXPECT_CALL(cm_, clusters()).WillOnce(Return(cluster_map));
+  EXPECT_THROW_WITH_MESSAGE(
+      subscriptionFromConfigSource(config), EnvoyException,
+      "envoy::api::v2::core::ConfigSource::GRPC must have a single gRPC service specified");
```

Comment:
I don't think you need to change anything now that we have these tests, but I would probably approach testing by doing full branch coverage in a dedicated unit test for `Utility::checkApiConfigSourceNames()` and then just do specific point tests in `SubscriptionFactoryTest` for where behavior is more complex. I think what you have is fine though.

Reply:
This is a good catch. The coverage report actually shows a hole in `utility.cc` for a REST API with defined gRPC services. I won't refactor this entirely but I am going to add coverage for that case.

Reply:
Done in bb137e7.